File: app/src/main/java/com/hussien/qouty/utilities/ActionsUtils.java
package com.hussien.qouty.utilities;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Typeface;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.core.content.FileProvider;

import com.hussien.qouty.AppExecutor;
import com.hussien.qouty.data.Quote;
import com.hussien.qouty.data.QuotesRepository;
import com.hussien.qouty.databinding.QuoteLayoutBinding;

import java.io.File;
import java.io.FileOutputStream;

public class ActionsUtils {
    private static final String TAG = "ActionsUtils";
    public static void updateQuoteLoved(Context context, Quote quote) {
        quote.setLoved(!quote.isLoved());
        QuotesRepository quotesRepository = QuotesRepository.getRepo_instance(context);

        Runnable runnable = () -> quotesRepository.updateQuote(quote);
        AppExecutor.getInstance().getDiskIo().execute(runnable);
    }

    public static void shareText(Context context, Quote quote) {
        String shareContent = getContentTextToShare(quote);
        Intent shareIntent = new Intent();
        shareIntent.setAction(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT,  shareContent);

        Intent chooser = Intent.createChooser(shareIntent, "Share via");
        chooser.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(chooser);
    }



    public static String getContentTextToShare(Quote quote){
        return '\u201D'+ quote.getText() +'\u201C' + "\n" + quote.getAuthor();
    }

    public static void copyText(Context context, Quote quote) {
        String shareContent = ActionsUtils.getContentTextToShare(quote);
        ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText(shareContent, shareContent);
        clipboard.setPrimaryClip(clip);
        Toast.makeText(context, "Copied!", Toast.LENGTH_SHORT).show();
    }

    public static void shareImg(Context context, Quote quote) {
        sendImgIntent(context, generateQuoteLayout(context, quote));
    }

    public static Bitmap generateQuoteLayout(Context context, Quote quote) {
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();

        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(displayMetrics.widthPixels,
                LinearLayout.LayoutParams.WRAP_CONTENT);

        QuoteLayoutBinding quoteLayoutBinding = QuoteLayoutBinding.inflate(LayoutInflater.from(context));
        quoteLayoutBinding.getRoot().setBackgroundColor(Color.TRANSPARENT);
        quoteLayoutBinding.getRoot().setLayoutParams(params);

        quoteLayoutBinding.quoteText.setLayoutParams(params);

        quoteLayoutBinding.quoteText.setText(quote.getText());
        quoteLayoutBinding.quoteAuthor.setText(quote.getAuthor());
        Typeface typeface = SharedPreferenceUtils.getUserChoiceFontFamily(context);
        quoteLayoutBinding.quoteText.setTypeface(typeface);
        quoteLayoutBinding.quoteAuthor.setTypeface(typeface);

        View view = quoteLayoutBinding.getRoot();
        view.measure(displayMetrics.widthPixels, displayMetrics.heightPixels);
        Bitmap bitmap = Bitmap.createBitmap(view.getMeasuredWidth(), view.getMeasuredHeight(),
                Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        view.layout(0, 0, view.getMeasuredWidth(), view.getMeasuredHeight());
        view.draw(canvas);
        return bitmap;
    }

    private static void sendImgIntent(Context context, Bitmap bitmap) {
        try {
            File outputFile = new File(context.getExternalFilesDir(null), "quote.png");
            FileOutputStream fileOutputStream = new FileOutputStream(outputFile);

            bitmap.compress(Bitmap.CompressFormat.PNG, 100, fileOutputStream);

            fileOutputStream.flush();
            fileOutputStream.close();

            Uri uri = FileProvider.getUriForFile(context,
                    context.getApplicationContext().getPackageName() + ".provider",
                    outputFile);

            Intent intent = new Intent();
            intent.setAction(Intent.ACTION_SEND);
            intent.setDataAndType(uri, "image/*");
            intent.putExtra(Intent.EXTRA_STREAM, uri);

            Intent chooser = Intent.createChooser(intent, "Share Image Via");
            chooser.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(chooser);
        }catch (Exception e){
            Log.e(TAG, "sendImgIntent: " + e.toString());
        }
    }
}

File: app/src/main/java/com/hussien/qouty/utilities/ActionsUtils.kt
package com.hussien.qouty.utilities

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Typeface
import android.net.Uri
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import androidx.core.content.FileProvider
import com.hussien.qouty.annotations.OpenForTesting
import com.hussien.qouty.data.QuotesRepository
import com.hussien.qouty.databinding.QuoteLayoutBinding
import com.hussien.qouty.di.ApplicationScope
import com.hussien.qouty.di.UIDispatcher
import com.hussien.qouty.ext.toBitmap
import com.hussien.qouty.models.Quote
import com.hussien.qouty.ui.settings.SettingsDataStoreManager
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Actions utils.
 *
 * the logic of execution the user click on any action of the quotes
 */
@Singleton
@OpenForTesting
class ActionsUtils @Inject constructor(
    @ApplicationContext private val context: Context,
    @ApplicationScope private val scope: CoroutineScope,
    @UIDispatcher private val uiDispatcher: MainCoroutineDispatcher = Dispatchers.Main,
    private val repository: QuotesRepository,
    private val settingsDataStoreManager: SettingsDataStoreManager
) {

    fun addQuoteToFavourites(quoteId: Long) {
        scope.launch {
            repository.addToQuoteFavourite(quoteId)
        }
    }

    fun shareText(quoteId: Long) {
        scope.launch {
            val content = getQuoteContentText(quoteId)
            val shareIntent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_TEXT, content)
            }
            val chooser: Intent = Intent.createChooser(shareIntent, "Share via")
            chooser.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(chooser)
        }
    }

    private suspend fun getQuoteContentText(quoteId: Long): String {
        val quote = repository.getQuote(quoteId)
        return quote.content
    }

    fun copyText(quoteId: Long) {
        scope.launch {
            val content = getQuoteContentText(quoteId)
            val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText(content, content)
            clipboard.setPrimaryClip(clip)
            withContext(uiDispatcher) {
                Toast.makeText(context, "Copied!", Toast.LENGTH_SHORT).show()
            }
        }
    }

    fun shareImg(quoteId: Long) {
        scope.launch {
            // get the quote
            val quote = repository.getQuote(quoteId)
            sendImgIntent(generateQuoteLayout(quote).toBitmap())
        }
    }

    suspend fun generateQuoteLayout(quote: Quote): View {
        // get the current width
        val displayMetrics: DisplayMetrics = context.resources.displayMetrics
        val orientation = context.resources.configuration.orientation
        val width = if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            displayMetrics.widthPixels
        } else {
            displayMetrics.heightPixels
        }
        val params = LinearLayout.LayoutParams(
            width,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )
        // generate the view
        val quoteLayoutBinding = withContext(uiDispatcher) {
            QuoteLayoutBinding.inflate(LayoutInflater.from(context))
        }
        quoteLayoutBinding.quote = quote

        quoteLayoutBinding.root.layoutParams = params
        quoteLayoutBinding.quoteText.layoutParams = params

        val fontName = settingsDataStoreManager.settingsPreferencesFlow.map { it.font }.first()
        val typeface: Typeface = Typeface.createFromAsset(context.assets, "fonts/$fontName")
        quoteLayoutBinding.quoteText.typeface = typeface
        quoteLayoutBinding.quoteAuthor.typeface = typeface

        quoteLayoutBinding.executePendingBindings()
        return quoteLayoutBinding.root
    }

    private fun sendImgIntent(bitmap: Bitmap) {
        try {
            val outputFile = File(context.getExternalFilesDir(null), "quote.png")
            FileOutputStream(outputFile).use {
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, it)
            }
            val uri: Uri = FileProvider.getUriForFile(
                context,
                context.applicationContext.packageName + ".provider",
                outputFile
            )
            val intent = Intent(Intent.ACTION_SEND).apply {
                setDataAndType(uri, "image/*")
                putExtra(Intent.EXTRA_STREAM, uri)
            }
            val chooser: Intent = Intent.createChooser(intent, "Share Image Via")
            chooser.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(chooser)
        } catch (e: Exception) {
            Timber.e(e)
        }
    }
}